### PR TITLE
Make puma log silence completely

### PR DESCRIPTION
**Pull request by ta1kt0me**
Puma log is shown that puma server starts. Using `action_dispatch_system_test_case` load hook can stop its output 

Before
```
$ bin/rspec spec/system/tasks_spec.rb
Capybara starting Puma...
* Version 3.12.0 , codename: Llamas in Pajamas
* Min threads: 0, max threads: 4
* Listening on tcp://127.0.0.1:51998
.

Finished in 1.39 seconds (files took 0.91103 seconds to load)
1 example, 0 failures
```

After
```
$ bin/rspec spec/system/tasks_spec.rb
.

Finished in 1.22 seconds (files took 0.84185 seconds to load)
1 example, 0 failures
```

### CODE REVIEW

**Comment by benoittgt** [comment]
Hello @ta1kt0me 

Thanks a lot for your patch. By any chance do you think you can provide a test for this?

I will look at the CI (Puma requir Ruby 2.2+ minimum now).

**Comment by ta1kt0me** [comment]
Thanks for your comment, I'll try it.

**Comment by JonRowe** [comment]
Honestly I'll be happy to merge this when green (depends on #2013)

**Comment by benoittgt** [comment]
CI is now green. Do you think you can rebase a re-push? I will look at this, this afternoon. 👋

**Comment by benoittgt** [comment]
Hello

It seems your test is broken. What I did instead is:
- clone the repo of https://github.com/rspec/rspec-rails/issues/1897#issuecomment-401015183
- Specify your branch and other rspec ones https://github.com/benoittgt/puma_in_rspec_output/commit/df6d432da8e72a3311638396819c9d5c55a11d2b
- Then when I run `rspec` I don't see anymore the puma log.

The test fails on:

```
$ bundle exec cucumber features/system_specs/system_specs.feature:38                                                                                      [20:08:02]
/Users/bti/code/bundle/ruby/2.5.0/gems/gherkin-2.12.2/lib/gherkin/formatter/filter_formatter.rb:105: warning: constant ::Fixnum is deprecated
Using the default profile...
..F--

(::) failed steps (::)

Exit status was 1 but expected it to be 0. Output:


Randomized with seed 7465

Widget management
  enables me to create widgets (FAILED - 1)

Failures:

  1) Widget management enables me to create widgets
     Failure/Error: raise ActionController::RoutingError, "No route matches [#{env['REQUEST_METHOD']}] #{env['PATH_INFO'].inspect}"

     ActionController::RoutingError:
       No route matches [GET] "/stylesheets/application.css"



     # /Users/bti/code/bundle/ruby/2.5.0/gems/railties-5.2.0/lib/rails/rack/logger.rb:38:in `call_app'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/railties-5.2.0/lib/rails/rack/logger.rb:26:in `block in call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/railties-5.2.0/lib/rails/rack/logger.rb:26:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/method_override.rb:22:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/runtime.rb:22:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/sendfile.rb:111:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/railties-5.2.0/lib/rails/engine.rb:524:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/urlmap.rb:68:in `block in call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/urlmap.rb:53:in `each'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/urlmap.rb:53:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/rack-2.0.5/lib/rack/builder.rb:153:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/capybara-2.18.0/lib/capybara/server.rb:44:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/configuration.rb:225:in `call'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:658:in `handle_request'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:472:in `process_client'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:332:in `block in run'
     # /Users/bti/code/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/thread_pool.rb:133:in `block in spawn_thread'
     # ------------------
     # --- Caused by: ---
     # Capybara::CapybaraError:
     #   Your application server raised an error - It has been raised in your test code because Capybara.raise_server_errors == true
     #   /Users/bti/code/bundle/ruby/2.5.0/gems/capybara-2.18.0/lib/capybara/session.rb:145:in `raise_server_error!'

Top 1 slowest examples (1.56 seconds, 99.8% of total time):
  Widget management enables me to create widgets
    1.56 seconds ./spec/system/widget_system_spec.rb:8

Finished in 1.56 seconds (files took 1.22 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/system/widget_system_spec.rb:8 # Widget management enables me to create widgets

Randomized with seed 7465


 (RSpec::Expectations::ExpectationNotMetError)
features/system_specs/system_specs.feature:59:in `Then the exit status should be 0'

Failing Scenarios:
cucumber features/system_specs/system_specs.feature:38 # Scenario: System specs driven by selenium_chrome_headless

1 scenario (1 failed)
5 steps (1 failed, 2 skipped, 2 passed)
0m5.098s
```

So fix is good for me. I don't know how much we invest to fix the test.

**Comment by JonRowe** [comment]
chromedriver isn't available on travis without installing it, if theres a driver that works built in that would work, but I can see the puma output in the output...

**Comment by benoittgt** [comment]
Are we accepting this patch without the test?

**Comment by JonRowe** [comment]
I think we need some form of integration or smoke test here really

**Comment by benoittgt** [comment]
@ta1kt0me do you think you have a look. 

Otherwise I will look at it.

**Comment by ta1kt0me** [comment]
@benoittgt @JonRowe I'm sorry for replying so late and thank you for your advice. I found the way which resolves https://github.com/rspec/rspec-rails/pull/2011#issuecomment-406885744 's error, but I can not be sure it's good solution.

```diff
diff --git a/Rakefile b/Rakefile
index 6bb2e0c4..435e361f 100644
--- a/Rakefile
+++ b/Rakefile
@@ -60,7 +60,7 @@ namespace :generate do

       # Rails 4 cannot use a `rails` binstub generated by Bundler
       sh "rm -f #{bindir}/rails"
-      sh "bundle exec rails new ./tmp/example_app --no-rc --skip-javascript --skip-sprockets --skip-git --skip-test-unit --skip-listen --skip-bundle --template=example_app_generator/generate_app.rb"
+      sh "bundle exec rails new ./tmp/example_app --no-rc --skip-javascript --skip-git --skip-test-unit --skip-listen --skip-bundle --template=example_app_generator/generate_app.rb"

       in_example_app do
         sh "./travis_retry_bundle_install.sh 2>&1"
```

The above makes all spec in green in my local. Or if you have other better solution, it might be good.

**Comment by benoittgt** [comment]
I think the intent to skip sprocket was to make the app smaller. 

Do you think you can add a commit with this change to check the CI?

**Comment by ta1kt0me** [comment]
I've added the change.

**Comment by benoittgt** [comment]
It seems their is an issue with the webdriver.

```
             Unable to find chromedriver. Please download the server from http://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH. More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.
```

**Comment by JonRowe** [comment]
As per my earlier comment:

> chromedriver isn't available on travis without installing it, if theres a driver that works built in that would work, but I can see the puma output in the output...

:joy:

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,8 @@
 language: ruby
 
-sudo: false
+sudo: required
+addons:
+  chrome: stable
```
Comment (BobbyMcWho) [inline .travis.yml]:
```suggestion
  apt:
    packages:
      - chromium-chromedriver
```
https://github.com/travis-ci/travis-ci/issues/272#issuecomment-389598926

According to that, you also have to add: 
```
before_script:
  - ln -s /usr/lib/chromium-browser/chromedriver ~/bin/chromedriver
```

Hopefully that works, I haven't tried it out personally.

Comment (benoittgt) [reply .travis.yml]: @ta1kt0me Do you think you can try?

Thanks @BobbyMcWho for the suggestion

Comment (ta1kt0me) [reply .travis.yml]: @BobbyMcWho @benoittgt 
Thank you for advice. I'll try it.

File: lib/rspec/rails/example/system_example_group.rb [hunk 2]
```diff
@@ -75,6 +75,10 @@ def app
 
         attr_reader :driver
 
+        if ::Rails.version.to_f == 5.1
```
Comment (benoittgt) [inline lib/rspec/rails/example/system_example_group.rb]: Could you explain why?

**Comment by benoittgt** [comment]
CI seems to be better.

File: lib/rspec/rails/example/system_example_group.rb (hunk 2 above)
Comment (ta1kt0me) [reply lib/rspec/rails/example/system_example_group.rb]: The first implementation is lack of considering of compatibility. `action_dispatch_system_test_case` is available from Rails 5.2 . The change is needed to pass `System specs driven by selenium_chrome_headless` spec in Rails 5.1 .

Comment (JonRowe) [reply lib/rspec/rails/example/system_example_group.rb]: It was literally checking the method exists, then using it, can you explain the "lack of compatibility" I'm afraid I don't understand how "not calling the method if it doesn't exist" changes anything... Also note this need to work on 5.2 and beyond, so this check seems incorrect.

File: Rakefile [hunk 3]
```diff
@@ -60,7 +60,7 @@ namespace :generate do
 
       # Rails 4 cannot use a `rails` binstub generated by Bundler
       sh "rm -f #{bindir}/rails"
-      sh "bundle exec rails new ./tmp/example_app --no-rc --skip-javascript --skip-sprockets --skip-git --skip-test-unit --skip-listen --skip-bundle --template=example_app_generator/generate_app.rb"
+      sh "bundle exec rails new ./tmp/example_app --no-rc --skip-javascript --skip-git --skip-test-unit --skip-listen --skip-bundle --template=example_app_generator/generate_app.rb"
```
Comment (benoittgt) [inline Rakefile]: Could you explain why you changed this?

This has a cost. It seems that sprockets adds sass gem. On Ruby 1.9.3 it adds deprecation warning and breaks the CI

```
     Failure/Error:
       expect(capture_exec(custom_env, exec_script)).to eq(
         "#{::Rails.root}/spec/mailers/previews"
       )
     
       expected: "/home/travis/build/rspec/rspec-rails/tmp/example_app/spec/mailers/previews"
            got: #<struct CaptureExec io="DEPRECATION WARNING:\nSass 3.5 will no longer support Ruby 1.9.3.\nPlease up...ible.\n\n/home/travis/build/rspec/rspec-rails/tmp/example_app/spec/mailers/previews", exit_status=0>
```

I revert this line and test on Ruby 1.9.3 and 2.5.3 without issue. I didn't test on all rails version. But maybe it doesn't need to be committed?

Comment (JonRowe) [reply Rakefile]: Yep I'd like this reverted too

**Comment by pirj** [comment]
@ta1kt0me Can you please rebase? CI is more stable lately.

**Comment by pirj** [comment]
@ta1kt0me Would you like to wrap this up before we release 4.0?

**Comment by benoittgt** [comment]
I can handle it if needed

**Comment by pirj** [comment]
This would further confirm your already quite intense comeback!

**Pull request closed without merging**